### PR TITLE
MNT-18895: Extension of Behaviour Filter for smart folders 

### PR DESCRIPTION
**Pull request by oussamaMesseguem**
The MNT-18895 ticket has highlighted the fact that behaviour filter service did not have an extension when it comes to smart folders.

The issue had to do with the creation of renditions as children of documents, when a document hasn't got any rendition yet, and the rendition is created as a new node from a smart folder, it is only the virtual node that gets its aspect cm:auditable disbaled and not the actual node linked to it. The cm:auditable is responsible for certain updates of node's properties, in particular the properties cm:modifier and cm:modified.

This PR proposes an extension for behaviour filter service as it is the case already for other services (node, coci, lock, fileFolder..)

### CODE REVIEW

File: src/main/java/org/alfresco/repo/policy/BehaviourFilter.java [hunk 1]
```diff
@@ -67,23 +68,27 @@
     /**
      * @deprecated Since 4.0 use {@link #enableBehaviour(NodeRef)}
      */
+    @Auditable(parameters = { "nodeRef" })
```
Comment (alandavis) [inline src/main/java/org/alfresco/repo/policy/BehaviourFilter.java]: Are you sure you need to @Auditable? I would not expect that to be the case. Can you try removing this and running the tests again.

Comment (oussamaMesseguem) [reply src/main/java/org/alfresco/repo/policy/BehaviourFilter.java]: Indeed @Auditable is not required, has been removed in [commit](https://github.com/Alfresco/alfresco-repository/pull/791/commits/d798ce02ae0e7c0ceb2a8eb6bf207bfbf8c72093).

File: src/main/java/org/alfresco/repo/policy/BehaviourFilterImpl.java [hunk 2]
```diff
@@ -62,7 +69,14 @@
     
     private DictionaryService dictionaryService;
     private TenantService tenantService;
-    
+
+    private final ExtendedTrait<BehaviourFilterTrait> behaviourFilterTrait;
+
+    public BehaviourFilterImpl() {
```
[the remark this answers is not in the document]
Comment (oussamaMesseguem) [reply src/main/java/org/alfresco/repo/policy/BehaviourFilterImpl.java]: Done in [commit](https://github.com/Alfresco/alfresco-repository/commit/5d276674b5ab12de9c126abbd65399f8f46cd12e).

File: src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class VirtualBehaviourFilterExtensionTest extends VirtualizationIntegrationTest
+{
+
```
[the remark this answers is not in the document]
Comment (oussamaMesseguem) [reply src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java]: Done in [commit](https://github.com/Alfresco/alfresco-repository/commit/02f459602c405580210601aa1fcb8d4757fec64e)

File: src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                ContentModel.ASSOC_CONTAINS,
+                "Node1");
+
+        virtualNodeRef = createContent(
+                nodeRef,
+                "actualContentName",
+                "0",
+                MimetypeMap.MIMETYPE_TEXT_PLAIN,
+                "UTF-8").getChildRef();
+
```
[the remark this answers is not in the document]
Comment (oussamaMesseguem) [reply src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java]: Done in [commit](https://github.com/Alfresco/alfresco-repository/commit/02f459602c405580210601aa1fcb8d4757fec64e)

File: src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                behaviourFilter.isEnabled(virtualNodeRef, ContentModel.ASPECT_AUDITABLE));
+
+        NodeRef actualNodeRef = smartStore.materialize(Reference.fromNodeRef(virtualNodeRef));
+
+        // Disable the aspect auditable using the virtual node
+        behaviourFilter.disableBehaviour(virtualNodeRef, ContentModel.ASPECT_AUDITABLE);
+
+        assertFalse("The auditable aspect for the actual node must not be enable since it has been disabled",
+                behaviourFilter.isEnabled(actualNodeRef, ContentModel.ASPECT_AUDITABLE));
+
```
[the remark this answers is not in the document]
Comment (oussamaMesseguem) [reply src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java]: Done in [commit](https://github.com/Alfresco/alfresco-repository/commit/02f459602c405580210601aa1fcb8d4757fec64e)

File: src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                behaviourFilter.isEnabled(actualNodeRef));
+
+        // Disabling the aspect auditable for the virtual node
+        behaviourFilter.disableBehaviour(virtualNodeRef);
+
+        assertFalse("The actual node must not be enable since it has been disabled",
+                behaviourFilter.isEnabled(actualNodeRef));
+
+    }
+
```
[the remark this answers is not in the document]
Comment (oussamaMesseguem) [reply src/test/java/org/alfresco/repo/virtual/bundle/VirtualBehaviourFilterExtensionTest.java]: Done in [commit](https://github.com/Alfresco/alfresco-repository/commit/02f459602c405580210601aa1fcb8d4757fec64e)

File: src/test/java/org/alfresco/AppContextExtraTestSuite.java [hunk 7]
```diff
@@ -139,6 +139,7 @@
     org.alfresco.repo.virtual.config.NodeRefPathExpressionTest.class,
     org.alfresco.repo.virtual.template.TemplateFilingRuleTest.class,
     org.alfresco.repo.virtual.bundle.FileInfoPropsComparatorTest.class,
+        org.alfresco.repo.virtual.bundle.VirtualBehaviourFilterExtensionTest.class,
```
[the remark this answers is not in the document]
Comment (oussamaMesseguem) [reply src/test/java/org/alfresco/AppContextExtraTestSuite.java]: Done in [commit](https://github.com/Alfresco/alfresco-repository/commit/0d50d1b92b52fdc1998f3a53f317dbf7945427b2)

**Pull request merged**